# Logestic Regression Model - SFS

In [34]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sqlalchemy import create_engine
from config import db_password
import matplotlib.pyplot as plt
import seaborn as sns
import math

In [35]:
db_string = f"postgresql://postgres:{db_password}@127.0.0.1:5432/wnba_champs"
engine = create_engine(db_string)

In [36]:
stats_df = pd.read_sql("select * from \"stats\"", db_string);
stats_df.head()

index  Position   Season   Tm   Age   W   L  Win_pct  PW  PL  ...   3PAr  \
0      0         1  1997-98  HOC  28.8  18  10   0.6429  22   6  ...  0.292   
1      1         0  1997-98  PHM  27.9  16  12   0.5714  20   8  ...  0.247   
2      2         0  1997-98  NYL  27.5  17  11   0.6071  17  11  ...  0.198   
3      3         0  1997-98  LAS  25.5  14  14   0.5000  17  11  ...  0.136   
4      4         0  1997-98  CLR  28.1  15  13   0.5357  16  12  ...  0.108   

   TS_pct  eFG_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TOV_pct  \
0   0.523    0.473     20.4     36.2   0.266        0.444         22.3   
1   0.473    0.411     18.8     33.2   0.274        0.441         24.2   
2   0.482    0.439     21.3     34.6   0.236        0.422         22.7   
3   0.504    0.464     20.4     32.7   0.235        0.423         19.4   
4   0.515    0.464     21.6     32.2   0.276        0.449         20.2   

   opp_DRB%  Opp_FT_FGA  
0      64.7       0.236  
1      68.5       0.245  
2      64.7       0.246  
3      67.9       0.263  
4      67.1       0.219  

[5 rows x 28 columns]

In [37]:
test_df= pd.read_sql("select * from \"cy_stats\"", db_string)
test_df.head()

index  Position   Season   Tm   Age   W   L  Win_pct  PW  PL  ...   3PAr  \
0      0         0  2021-22  COS  28.3  26   6   0.8125  28   4  ...  0.298   
1      1         1  2021-22  LVA  26.4  24   8   0.7500  26   6  ...  0.192   
2      2         0  2021-22  SES  28.1  21  11   0.6563  22  10  ...  0.328   
3      3         0  2021-22  MLY  27.6  22  10   0.6875  21  11  ...  0.296   
4      4         0  2021-22  PHM  28.4  19  13   0.5938  19  13  ...  0.332   

   TS_pct  eFG_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TOV_pct  \
0   0.543    0.497     15.9     31.2   0.220        0.459         16.1   
1   0.559    0.508     12.6     21.3   0.255        0.464         12.5   
2   0.556    0.512     13.9     20.9   0.206        0.483         14.4   
3   0.553    0.508     15.8     22.3   0.218        0.476         14.6   
4   0.553    0.507     14.7     23.8   0.235        0.466         11.3   

   opp_DRB%  Opp_FT_FGA  
0      82.1       0.201  
1      80.0       0.165  
2      78.6       0.180  
3      79.7       0.194  
4      75.1       0.167  

[5 rows x 28 columns]

In [38]:
# Pull out Season, TM and Team_Name into DF
name_df = stats_df[['Season', 'Tm',]]
name_df.head()

Season   Tm
0  1997-98  HOC
1  1997-98  PHM
2  1997-98  NYL
3  1997-98  LAS
4  1997-98  CLR

In [39]:
#list(stats_df.columns)

In [40]:
all_features_df = stats_df
all_features_df.head()

index  Position   Season   Tm   Age   W   L  Win_pct  PW  PL  ...   3PAr  \
0      0         1  1997-98  HOC  28.8  18  10   0.6429  22   6  ...  0.292   
1      1         0  1997-98  PHM  27.9  16  12   0.5714  20   8  ...  0.247   
2      2         0  1997-98  NYL  27.5  17  11   0.6071  17  11  ...  0.198   
3      3         0  1997-98  LAS  25.5  14  14   0.5000  17  11  ...  0.136   
4      4         0  1997-98  CLR  28.1  15  13   0.5357  16  12  ...  0.108   

   TS_pct  eFG_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TOV_pct  \
0   0.523    0.473     20.4     36.2   0.266        0.444         22.3   
1   0.473    0.411     18.8     33.2   0.274        0.441         24.2   
2   0.482    0.439     21.3     34.6   0.236        0.422         22.7   
3   0.504    0.464     20.4     32.7   0.235        0.423         19.4   
4   0.515    0.464     21.6     32.2   0.276        0.449         20.2   

   opp_DRB%  Opp_FT_FGA  
0      64.7       0.236  
1      68.5       0.245  
2      64.7       0.246  
3      67.9       0.263  
4      67.1       0.219  

[5 rows x 28 columns]

In [41]:
stats_df.head()

index  Position   Season   Tm   Age   W   L  Win_pct  PW  PL  ...   3PAr  \
0      0         1  1997-98  HOC  28.8  18  10   0.6429  22   6  ...  0.292   
1      1         0  1997-98  PHM  27.9  16  12   0.5714  20   8  ...  0.247   
2      2         0  1997-98  NYL  27.5  17  11   0.6071  17  11  ...  0.198   
3      3         0  1997-98  LAS  25.5  14  14   0.5000  17  11  ...  0.136   
4      4         0  1997-98  CLR  28.1  15  13   0.5357  16  12  ...  0.108   

   TS_pct  eFG_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TOV_pct  \
0   0.523    0.473     20.4     36.2   0.266        0.444         22.3   
1   0.473    0.411     18.8     33.2   0.274        0.441         24.2   
2   0.482    0.439     21.3     34.6   0.236        0.422         22.7   
3   0.504    0.464     20.4     32.7   0.235        0.423         19.4   
4   0.515    0.464     21.6     32.2   0.276        0.449         20.2   

   opp_DRB%  Opp_FT_FGA  
0      64.7       0.236  
1      68.5       0.245  
2      64.7       0.246  
3      67.9       0.263  
4      67.1       0.219  

[5 rows x 28 columns]

In [42]:
# Drop un needed columns
all_features_df.drop(['index',
 'Season',
 'Tm',
 'W',
 'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 #'Pace',
 #'ORtg',
 #'DRtg',
 #'eFG_pct',
 #'TS_pct',
 #'TOV_pct',
 #'ORB_pct',
 #'FT_FGA',
 #'Opp_eFG_pct',
 #'Opp_TS_pct',
 #'Opp_TOV_pct',
 #'Opp_ORB_pct',
 #'Opp_FT_FGA'
              ], axis=1, inplace=True)

In [43]:
all_features_df.head()

Position   Age  Win_pct  PW  PL   MOV   SOS   SRS  ORtg  DRtg  ...   3PAr  \
0         1  28.8   0.6429  22   6  6.32 -0.79  5.53  99.4  90.7  ...  0.292   
1         0  27.9   0.5714  20   8  4.00 -0.48  3.38  91.4  86.1  ...  0.247   
2         0  27.5   0.6071  17  11  2.39 -0.32  2.22  90.6  87.4  ...  0.198   
3         0  25.5   0.5000  17  11  2.21 -0.23  1.63  93.9  91.1  ...  0.136   
4         0  28.1   0.5357  16  12  1.79 -0.22  1.56  94.4  92.0  ...  0.108   

   TS_pct  eFG_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TOV_pct  \
0   0.523    0.473     20.4     36.2   0.266        0.444         22.3   
1   0.473    0.411     18.8     33.2   0.274        0.441         24.2   
2   0.482    0.439     21.3     34.6   0.236        0.422         22.7   
3   0.504    0.464     20.4     32.7   0.235        0.423         19.4   
4   0.515    0.464     21.6     32.2   0.276        0.449         20.2   

   opp_DRB%  Opp_FT_FGA  
0      64.7       0.236  
1      68.5       0.245  
2      64.7       0.246  
3      67.9       0.263  
4      67.1       0.219  

[5 rows x 23 columns]

In [44]:
# Remove Position target from features data
y = all_features_df.Position
X = all_features_df.drop(columns=["Position"])

# Split training/test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)


In [45]:
feature_names = X.columns

In [46]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Scale the data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

In [47]:
# Create a logistic regression model
classifier_AF = LogisticRegression(max_iter=1000)
classifier_AF

LogisticRegression(max_iter=1000)

In [48]:
# Fit (train) our model by using the training data
classifier_AF.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

In [49]:
# Validate the model by using the test data
print(f"Logistic Regression Training Data Score: {classifier_AF.score(X_train_scaled, y_train)}")
print(f"Logistic Regression Testing Data Score: {classifier_AF.score(X_test_scaled, y_test)}")

Logistic Regression Training Data Score: 0.9282700421940928
Logistic Regression Testing Data Score: 0.9240506329113924


All features score = 0.92405

Find accuracy using SFM(SelectFromModel)

In [50]:
# https://towardsdatascience.com/5-feature-selection-method-from-scikit-learn-you-should-know-ed4d116e4172
from sklearn.feature_selection import SelectFromModel
smf=SelectFromModel(classifier_AF)
smf.fit(X_train, y_train)
X.columns[smf.get_support()]

Index(['Win_pct', 'PW', 'MOV', 'SOS', 'SRS', 'ORtg', 'DRtg', 'NRtg',
       'TOV_pct'],
      dtype='object')

In [51]:
# Drop un needed columns
stats_df.drop([#'index',
 #'Season',
 #'Tm',
 #'W',
 #'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 'Pace',
 #'ORtg',
 #'DRtg',
 'eFG_pct',
 'TS_pct',
 'TOV_pct',
 'ORB_pct',
 'FT_FGA',
 'Opp_eFG_pct',
 'Opp_TOV_pct',
 'Opp_FT_FGA'
              ], axis=1, inplace=True)
stats_df.head()

Position   Age  Win_pct  PW  PL   MOV   SOS   SRS  ORtg  DRtg  NRtg    FTr  \
0         1  28.8   0.6429  22   6  6.32 -0.79  5.53  99.4  90.7   8.7  0.357   
1         0  27.9   0.5714  20   8  4.00 -0.48  3.38  91.4  86.1   5.3  0.359   
2         0  27.5   0.6071  17  11  2.39 -0.32  2.22  90.6  87.4   3.2  0.354   
3         0  25.5   0.5000  17  11  2.21 -0.23  1.63  93.9  91.1   2.8  0.348   
4         0  28.1   0.5357  16  12  1.79 -0.22  1.56  94.4  92.0   2.4  0.382   

    3PAr  opp_DRB%  
0  0.292      64.7  
1  0.247      68.5  
2  0.198      64.7  
3  0.136      67.9  
4  0.108      67.1

In [52]:
# Remove Position target from features data
y = stats_df.Position
X = stats_df.drop(columns=["Position"])

# Split training/test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

In [53]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Scale the data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

In [54]:
# Create a logistic regression model
classifier_SFM = LogisticRegression(max_iter=1000)
classifier_SFM

LogisticRegression(max_iter=1000)

In [55]:
# Fit (train) our model by using the training data
classifier_SFM.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

In [56]:
# Validate the model by using the test data
print(f"Logistic Regression with SFM features only Training Data Score: {classifier_SFM.score(X_train_scaled, y_train)}")
print(f"Logistic Regression with SFM features only Testing Data Score: {classifier_SFM.score(X_test_scaled, y_test)}")

Logistic Regression with SFM features only Training Data Score: 0.9240506329113924
Logistic Regression with SFM features only Testing Data Score: 0.9240506329113924


SFS feature score = 0.92405

In [57]:
# Pull out Season, TM and Team_Name into DF
test_name_df = test_df[['Season', 'Tm']]
test_name_df.head()

Season   Tm
0  2021-22  COS
1  2021-22  LVA
2  2021-22  SES
3  2021-22  MLY
4  2021-22  PHM

In [58]:
test_df_pos= test_df['Position']
test_df.drop(['index',
 'Position',
 'Season',
 'Tm',
 'W',
 'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 'Pace',
 #'ORtg',
 #'DRtg',
 'eFG_pct',
 'TS_pct',
 'TOV_pct',
 'ORB_pct',
 'FT_FGA',
 'Opp_eFG_pct',
 'Opp_TOV_pct',
 'Opp_FT_FGA'
             ], axis=1, inplace=True)
test_df.head()

Age  Win_pct  PW  PL   MOV   SOS   SRS   ORtg   DRtg  NRtg    FTr   3PAr  \
0  28.3   0.8125  28   4  9.81 -0.95  8.86  106.9   93.7  13.2  0.270  0.298   
1  26.4   0.7500  26   6  9.06 -0.62  8.44  109.1   98.0  11.1  0.312  0.192   
2  28.1   0.6563  22  10  4.44 -0.63  3.80  106.3  100.7   5.6  0.246  0.328   
3  27.6   0.6875  21  11  4.00 -0.37  3.63  104.2   99.1   5.1  0.260  0.296   
4  28.4   0.5938  19  13  2.56 -0.12  2.45  106.5  103.2   3.3  0.295  0.332   

   opp_DRB%  
0      82.1  
1      80.0  
2      78.6  
3      79.7  
4      75.1

In [59]:
df_X_test_scaled = scaler.fit_transform(test_df)

In [60]:
df_test_df=classifier_SFM.predict(df_X_test_scaled)
df_test_df

array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [61]:
cy_test_df = pd.DataFrame(df_test_df, columns = ['Position'])
cy_test_df.head()

Position
0         0
1         1
2         0
3         0
4         0

In [62]:
joined_df = test_name_df.join(cy_test_df)
joined_df

Season   Tm  Position
0   2021-22  COS         0
1   2021-22  LVA         1
2   2021-22  SES         0
3   2021-22  MLY         0
4   2021-22  PHM         0
5   2021-22  CSK         0
6   2021-22  DAW         0
7   2021-22  WAM         0
8   2021-22  LAS         0
9   2021-22  ATD         0
10  2021-22  NYL         0
11  2021-22  INF         0

In [64]:
classifier_SFM.predict_proba(df_X_test_scaled)

array([[5.86435121e-01, 4.13564879e-01],
       [4.71403060e-01, 5.28596940e-01],
       [8.63567536e-01, 1.36432464e-01],
       [8.94698828e-01, 1.05301172e-01],
       [8.86292149e-01, 1.13707851e-01],
       [9.68651891e-01, 3.13481093e-02],
       [9.87122993e-01, 1.28770072e-02],
       [9.94185388e-01, 5.81461215e-03],
       [9.98311335e-01, 1.68866503e-03],
       [9.97859135e-01, 2.14086481e-03],
       [9.98688768e-01, 1.31123223e-03],
       [9.99113839e-01, 8.86160529e-04]])